src/rgw/rgw_notify.cc

```diff
@@ -597,9 +597,59 @@ int remove_persistent_topic(const std::string& topic_name, optional_yield y) {
   return s_manager->remove_persistent_topic(topic_name, y);
 }
 
+rgw::sal::RGWObject* get_object_with_atttributes(const req_state* s, rgw::sal::RGWObject* obj) {
+  // in case of copy obj, the tags and metadata are taken from source
+  const auto src_obj = s->src_object ? s->src_object.get() : obj;
+  if (src_obj->get_attrs().empty()) {
+    if (!src_obj->get_bucket()) {
+      src_obj->set_bucket(s->bucket.get());
+    }
+    if (src_obj->get_obj_attrs(s->obj_ctx, s->yield) < 0) {
+      return nullptr;
+    }
+  }
+  return src_obj;
+}
+
+void metadata_from_attributes(const req_state* s, rgw::sal::RGWObject* obj, KeyValueList& metadata) {
+  const auto src_obj = get_object_with_atttributes(s, obj);
+  if (!src_obj) {
+    return;
+  }
+  for (auto& attr : src_obj->get_attrs()) {
+    if (boost::algorithm::starts_with(attr.first, RGW_ATTR_META_PREFIX)) {
+      std::string_view key(attr.first);
+      key.remove_prefix(sizeof(RGW_ATTR_PREFIX)-1);
+      // we want to pass a null terminated version
+      // of the bufferlist, hence "to_str().c_str()"
+      metadata.emplace(key, attr.second.to_str().c_str());
+    }
+  }
+}
+
+void tags_from_attributes(const req_state* s, rgw::sal::RGWObject* obj, KeyValueList& tags) {
+  const auto src_obj = get_object_with_atttributes(s, obj);
+  if (!src_obj) {
+    return;
+  }
+  const auto& attrs = src_obj->get_attrs();
+  const auto attr_iter = attrs.find(RGW_ATTR_TAGS);
+  if (attr_iter != attrs.end()) {
+    auto bliter = attr_iter->second.cbegin();
+    RGWObjTags obj_tags;
+    try {
+      ::decode(obj_tags, bliter);
+    } catch(buffer::error&) {
+      // not able to decode tags
+      return;
+    }
+    tags = std::move(obj_tags.get_tags());
+  }
+}
+
 // populate record from request
 void populate_record_from_request(const req_state *s, 
```

Comment:
is "record" part of notification terminology?  what exactly is it?

Reply:
the term "record" is taken from the structure here: https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html :
```
{  
   "Records":[  
      {  
         "eventVersion":"2.2",
         "eventSource":"aws:s3",
         "awsRegion":"us-west-2",
...
```
however, we could have used the term "event" or the term "notification" (also used in the AWS doc).

Reply:
ok, I think match and record are a bit free of context for me, but whatevs

Reply:
will change to "event" as this is what AWS uses in their doc.
"record" is just the term inside the xml